### PR TITLE
Feat: Block instance replication of itself and it's own replicas

### LINKED ISSUE #1484: Don't allow to slaveof itself or its replica
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-kvrocks/issues) and found no similar issues.

### Motivation

Currently, we didn't check if the master is itself or in its replica set, so it may cause a replication loop like the below:

```
127.0.0.1:6666> SLAVEOF 127.0.0.1 6666
OK
127.0.0.1:6666> info replication
# Replication
role:slave
master_host:127.0.0.1
master_port:6666
master_link_status:up
master_sync_unrecoverable_error:no
master_sync_in_progress:0
master_last_io_seconds_ago:2
slave_repl_offset:0
slave_priority:100
connected_slaves:1
slave0:ip=127.0.0.1,port=6666,offset=0,lag=0
master_repl_offset:0
```

### Solution

Check the master host/port before sending the command.

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by uds5501**
WIP: A patch to block an instance from replicating itself and it's own replicas

Fixes: #1484

### CODE REVIEW

**Comment by git-hulk** [comment]
Thanks @uds5501, you need to fix the clang-tidy error: https://github.com/apache/incubator-kvrocks/actions/runs/5199313230/jobs/9376587287?pr=1488#step:8:1166 to pass the CI. And can you add some Go tests to make sure it works?

**Comment by uds5501** [comment]
@git-hulk , added integration test and clang-tidy error

**Comment by git-hulk** [comment]
> @git-hulk , added integration test and clang-tidy error

Thanks for your efforts

File: src/server/server.cc [hunk 1]
```diff
@@ -1751,3 +1751,16 @@ void Server::ResetWatchedKeys(redis::Connection *conn) {
...
+  slave_threads_mu_.lock();
+  for (const auto &slave : slave_threads_) {
+    if (slave->IsStopped()) continue;
+    std::pair<std::string, int> host_port_pair = {slave->GetConn()->GetAnnounceIP(),
+                                                  slave->GetConn()->GetListeningPort()};
+    result.emplace_back(host_port_pair);
+  }
+  slave_threads_mu_.unlock();
+  return result;
+}
```
Comment (PragmaTwice) [inline src/server/server.cc]:
```suggestion
}

```

Newline in the end of the file is expected.

**Review by git-hulk — APPROVED** [review]

**Review by torwig — APPROVED** [review]
LGTM

**Comment by git-hulk** [comment]
Thanks all, merging...

**Comment by git-hulk** [comment]
@uds5501 Thanks for your contribution.

**Pull request merged**